iio: adc: ad7292: Add analog input channels and suport for voltage regulator

marceloschmitt1 opened the pull request:
This pull request adds IIO channels for the eight multiplexed analog input channels included in the AD7292 part. It also adds a voltage regulator to handle its internal voltage reference.

commodo commented on drivers/iio/adc/ad7292.c:
```diff
@@ -86,6 +119,10 @@ static int ad7292_read_raw(struct iio_dev *indio_dev,
 			*val = 25;
 			ret = IIO_VAL_INT;
 			break;
+		case IIO_GENERIC_DATA:
```
please, don't use IIO_GENERIC_DATA
it's not upstream; it's used for some project that ADI had with Mathworks
also, the vendor ID is not something we typically add in a read_raw, so maybe just print it during probe (if you want it)

marceloschmitt1 replied:
OK, I'll just remove it along with the SPI commit. Since there are some changes to be made in previous commits, I'll focus on that first. Thanks.

commodo commented on drivers/iio/adc/ad7292.c:
```diff
@@ -42,6 +51,11 @@ struct ad7292_state {
 	_channel2, BIT(IIO_CHAN_INFO_RAW), BIT(IIO_CHAN_INFO_SCALE),	\
 		_sign, 10, 16, 6, IIO_LE)				\
 
+#define AD7292_VENDOR_CHANNEL()						\
```
typically, vendor channels are not defined

marceloschmitt1 replied:
OK, will remove it with the SPI commit.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -51,8 +65,26 @@ static const struct iio_chan_spec ad7292_channels[] = {
+		return ret;
+
+	return (be32_to_cpu(st->data.d32) >> shift);
+}
```
First, does this work? Has it been tested? The register sizes are either 8 or 16 , so I would rather store the DOUT data into something like a `d16 `variable instead of `d32`.

marceloschmitt1 replied:
I wanted to test it this weekend but didn't manage to do it. I still have to get a look at the spi-cpol and spi-cpha properties before testing. I'll also change AD7292 state to use d16 to hold DOUT data when I come back to it. Thanks.

stefpopa replied:
Normally, we try to test all the patches before creating a PR. Of course, if the change is not a functional one, for instance a white space fix, then a test is not required.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -51,8 +65,26 @@ static const struct iio_chan_spec ad7292_channels[] = {
 	AD7292_ADC_CHANNEL(0, 5, 0, 'u'),
 	AD7292_ADC_CHANNEL(0, 6, 0, 'u'),
 	AD7292_ADC_CHANNEL(0, 7, 0, 'u'),
+	AD7292_VENDOR_CHANNEL(),
```
The vendor ID, should not be a separate channel. Just read it and verify it during probe.

marceloschmitt1 replied:
OK, thanks.

stefpopa commented on Documentation/devicetree/bindings/iio/adc/adi,ad7292.yaml:
```diff
@@ -31,5 +38,6 @@ examples:
       compatible = "adi,ad7292";
       reg = <0>;
-      spi-max-frequency = <16000000>;
+      spi-max-frequency = <25000000>;
```
This change looks like it does not belong in this commit.

marceloschmitt1 replied:
Yeah, I noticed later that spi-max-freq for AD7292 was 25MHz and thought of correcting it along with the voltage regulator commit. Well, then I'll hold this change for a better opportunity.

stefpopa replied:
The change is not wrong, it just fits better in a separate commit.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -147,6 +157,25 @@ static int ad7292_probe(struct spi_device *spi)
+			return ret;
+
+		st->vref_mv = ret / 1000;
+	}
```
The ad7292 can use an external or internal vref. If the vref fixed regulator is present in the devicetree, then we we should use an external vref, otherwise the internal vref should be used.
It's a good thing that you used `devm_regulator_get_optional()`, however, we need to set st->vref_mv to 1250 when internal vref is used. Take a look at this example: https://github.com/analogdevicesinc/linux/blob/rpi-4.14.y_GSOC_2019/drivers/iio/adc/ad5940.c#L257
Another thing thing that we should take into account is the fact that ad7292 can use different ranges for the vref: 2 * vref and 4 * vref. The way this is accomplished is by using the IIO_CHAN_INFO_SCALE attribute, here's an example: https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/adc/ad7606.c#L264

marceloschmitt1 replied:
Right, will push these changes on the next PR version. Thanks.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -147,6 +157,25 @@ static int ad7292_probe(struct spi_device *spi)
+		st->vref_mv = ret / 1000;
+	}
+
+	ret = devm_add_action_or_reset(&spi->dev, ad7292_regulator_disable, st);
```
This function should be called right after calling `regulator_enable()`. The way it is done right now is that we might be calling `regulator_disable()` even if we we haven't enabled the regulator in the first place.

marceloschmitt1 replied:
Makes sense. Didn't realized that before. I'll pull it up inside the devm_regulator_get_optional if as it is in Song's driver. Thanks.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -15,6 +15,41 @@ struct ad7292_state {
+	AD7292_ADC_CHANNEL(0, 5, 0, 'u'),
+	AD7292_ADC_CHANNEL(0, 6, 0, 'u'),
+	AD7292_ADC_CHANNEL(0, 7, 0, 'u'),
+};
```
The ad7292 has 8 channels, by default these channels are configured as single-ended channels (unipolar), however, VIN0 and VIN1 channels can be configured as differential (bipolar), the rest of the channels are single-ended. 
The way we can deal with this case is by adding channel nodes, therefore we can configure each channel in a particular way. For now, let's add only the `diff-channels` property which should be available for the VIN-VIN1 pair.

marceloschmitt1 replied:
So, let me see if I get it right. First I'll refactor AD7292_ADC_CHANNEL macro to set all the ADC channels (but VIN0 and VIN1) as single-ended mode by default. For VIN0 and VIN1 there will be configurable channel nodes that will be customized at probe according to some diff-channels property present for VIN0 and VIN1 at devicetree. Is this device node set up similar to what is done at AD7124?
https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/adc/ad7124.c#L433

stefpopa replied:
That's right! Also, Song has done something more or less similar for ad5940, you can check that example as well.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -15,6 +15,41 @@ struct ad7292_state {
+#define AD7292_ADC_CHANNEL(_differential, _channel, _channel2, _sign)	\
+	AD7292_CHANNEL(IIO_VOLTAGE, _differential, 1, _channel,		\
+	_channel2, BIT(IIO_CHAN_INFO_RAW), BIT(IIO_CHAN_INFO_SCALE),	\
+		_sign, 10, 16, 6, IIO_LE)				\
```
This macro does too much stuff. Whether the channel is differential or not, should be taken from the devicetree. Also, if the channel is differential, then it should be signed, otherwise unsigned.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -15,6 +15,41 @@ struct ad7292_state {
 	struct spi_device *spi;
 };
 
+#define AD7292_CHANNEL(_type, _differential, _indexed, _channel,	\
```
This macro does too much stuff. Whether the channel is differential or not, should be taken from the devicetree. Also, if the channel is differential, then it should be signed, otherwise unsigned.
All channels should be indexed.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -15,6 +15,41 @@ struct ad7292_state {
+#define AD7292_CHANNEL(_type, _differential, _indexed, _channel,	\
+	_channel2, _info_mask_separate, _info_mask_shared_by_type,	\
+	_sign, _realbits, _storagebits, _shift, _endianness) {		\
+	.type = (_type),						\
```
This macro does too much stuff. Whether the channel is differential or not, should be taken from the devicetree. Also, if the channel is differential, then it should be signed, otherwise unsigned.
All channels should be indexed.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -15,6 +15,41 @@ struct ad7292_state {
+	.channel2 = (_channel2),					\
+	.info_mask_separate = (_info_mask_separate),			\
+	.info_mask_shared_by_type = (_info_mask_shared_by_type),	\
+	.scan_type = {							\
```
This field is used by the iio buffers for which there is yet no support in the driver. Let's add this later.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -24,7 +59,65 @@ static int ad7292_read_raw(struct iio_dev *indio_dev,
+			break;
+		}
+		break;
+	case IIO_CHAN_INFO_SCALE:
```
Let's add this later

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -24,7 +59,65 @@ static int ad7292_read_raw(struct iio_dev *indio_dev,
+				break;
+			}
+			break;
+		case IIO_TEMP:
```
let's add this later

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -24,7 +59,65 @@ static int ad7292_read_raw(struct iio_dev *indio_dev,
+				ret = IIO_VAL_INT;
+			}
+			break;
+		case IIO_TEMP:
```
let's add this later

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -24,7 +59,65 @@ static int ad7292_read_raw(struct iio_dev *indio_dev,
+	int ret = -EINVAL;
+
+	switch (info) {
+	case IIO_CHAN_INFO_RAW:
```
for now, let's only have something like this:
```
	switch (info) {
	case IIO_CHAN_INFO_RAW:
		*val = 1000;

		return IIO_VAL_INT;
```

marceloschmitt1 commented:
### Changelog v2
- refactored AD7292 channels macros
- removed switch and cases for some channels types at `ad7292_read_raw`
- added dtbinding doc to configure differential mode through `channelNode` property

commodo commented:
where possible, try to check the build;
some errors it catches are valid:
https://travis-ci.org/analogdevicesinc/linux/builds/556078427#L2013

stefpopa commented on Documentation/devicetree/bindings/iio/adc/adi,ad7292.yaml:
```diff
@@ -12,7 +12,9 @@ maintainers:
   Analog Devices AD7292 10-Bit Monitor and Control System with ADC, DACs,
   Temperature Sensor, and GPIOs
-  https://www.analog.com/en/products/ad7292.html
+
```
This commit looks good.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -15,6 +15,39 @@ struct ad7292_state {
+
+static const struct iio_chan_spec ad7292_channels[] = {
+	AD7292_ADC_CHANNEL(0, 'u'),
+	AD7292_ADC_CHANNEL(1, 'u'),
```
You need to get the diff-channel info from the devicetree for these channels.

marceloschmitt1 replied:
From what I understood from the AD7124 driver, it has only differential or pseudo-differential analog inputs (if I didn't misread its datasheet). So, I guess it is ok to handle the "other channels" with channel_configs however, AD7292 analog inputs are by default single-ended with the possibility of configuring only VIN0 and VIN1 as differential (I guess they would become a single input channel working if set in differential mode). So I need to manage channels at `probe` to build both single-ended and (possibly) a differential channel to be given to indio_dev. I'm working on it now.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -15,6 +15,39 @@ struct ad7292_state {
 	struct spi_device *spi;
 };
 
+#define AD7292_CHANNEL(_type, _indexed, _channel,			\
```
I don't think we will need this macro

marceloschmitt1 replied:
OK, removed it on V3. Thanks.

marceloschmitt1 commented:
### Changelog V3
- Added a channel_template for the analog input channels
- Removed AD7292_CHANNEL macros
- Added a `ad7292_of_parse_channel_config` to parse and config differential mode channels
- Changed `probe` to add the single-ended and differential mode analog channels

marceloschmitt1 commented:
### Changelog V4
- Refactored `ad7292_of_parse_channel_config` to parse and config differential mode channels
- Refactored `probe` to add the single-ended and (if set) the differential channels
- Updated dtoverlay to set an analog differential channel

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -11,10 +11,81 @@
+		.storagebits = 16,
+		.shift = 6,
+		.endianness = IIO_LE,
+	},
```
there is no need for the `scan_type `field just yet. We will use it in the future, when IIO buffers support will be added. So, for now we can remove it.

marceloschmitt1 replied:
OK. I'll add it later then.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -39,13 +118,12 @@ static const struct iio_info ad7292_info = {
 	struct ad7292_state *st;
 	struct iio_dev *indio_dev;
+	struct iio_chan_spec *chan;
+	int i;
```
nitpick: you can fit the `i `and `ret `variables on the same line: `int i, ret;`

marceloschmitt1 replied:
OK. The `i` variable is there just to help to create several `ad7292_channel_template` for the single-ended inputs. If we change to use macros for creating those channels, this variable won't be needed anymore.

mirceacaprioru commented on Documentation/devicetree/bindings/iio/adc/adi,ad7292.yaml:
```diff
@@ -26,10 +26,46 @@ required:
   - compatible
   - reg
 
+channelNode:
```
I don't think these it the way to describe channelNodes. Take a look at https://github.com/analogdevicesinc/linux/blob/3609f860c08c7a3e2c658cada3ccf8ed997b62a3/Documentation/devicetree/bindings/iio/adc/adi%2Cad7124.yaml#L65

marceloschmitt1 replied:
OK, I'll change the doc to comply with the conventional way of describing channel nodes. Thanks.

marceloschmitt1 commented:
### Changelog V5

- Removed iio_chan_spec templates
- Added macros for setting the analog channels (both differential and single-ended)
- Removed deviceNode from dtoverlay and binding doc
- Added diff-channels as an optional boolean property for the AD7292 device tree node
- Added spi-cpha as a required property for the AD7292 dt node to have correct readings
- Added a function to read from AD7292 internal registers through SPI
- Added a `dev_info` call to print the vendor ID at end of `probe` function

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -12,6 +12,11 @@
 #include <linux/iio/iio.h>
 
+/* AD7292 registers definition */
+#define AD7292_REG_VENDOR_ID		0x00
```
the vendor ID is 0x0018

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -151,6 +177,11 @@ static int ad7292_probe(struct spi_device *spi)
+	if (ret)
+		dev_info(&spi->dev, "%s adc registered, vendor id: %d\n",
+			 indio_dev->name, ret);
+
```
You should also check if the vendor ID is what is being expected: 0x0018

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -52,8 +57,29 @@ struct ad7292_state {
+		return ret;
+
+	return (be16_to_cpu(st->data.d16) >> shift);
+}
```
Hmm.. it seems that it's a bit more complicated. By taking a look at the datasheet, page 18 and 19, it seems that we need to treat three cases displayed in figures 36, 37 and 38:
1. Accessing a 16 bit register (which you treated above, but requires slight modifications)
2. Accessing an 8-Bit Subregister Within a Register Bank
3. Accessing a 16-Bit Subregister Within a Register Bank
My suggestion is to have two spi_reg_read() functions: one that deals with the first case and a second which deals with the remaining two cases. 
The first function, doesn't need to take the length parameter, since the write and read lengths are known: we need to write 1 byte and read 2.
For the second function, we know that we need to write 2 bytes, but can read 1 byte (figure 37) or read 2 bytes (figure 38). Therefore a length parameter will be needed.

marceloschmitt1 commented:
### Changelog V6

- Added check for the correct vendor ID at `probe`
- Remove `len` parameter from `ad7292_spi_reg_read`
- Created the `ad7292_spi_subreg_read` function the read from subregisters

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -151,6 +191,14 @@ static int ad7292_probe(struct spi_device *spi)
+	ret = ad7292_spi_reg_read(st, AD7292_REG_VENDOR_ID);
+	if (ret != ADI_VENDOR_ID) {
+		dev_err(&spi->dev, "SPI communication not working properly.\n");
+		return -ECOMM;
```
Typically, in this case we print something like `"Wrong vendor id %x\n", ret`, and then return` -EINVAL`.

marceloschmitt1 replied:
OK, I'll change the message and return as suggested. Thanks.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -151,6 +191,14 @@ static int ad7292_probe(struct spi_device *spi)
+		return -ECOMM;
+	}
+
+	dev_info(&spi->dev, "Analog Devices %s ADC found\n", indio_dev->name);
```
I would remove this `dev_info()`

marceloschmitt1 replied:
OK. I'll do so.

marceloschmitt1 commented:
### Changelog 7

- Changed error message and return value when vendor ID is not correct.
- Removed `dev_info()` message at end of `probe`.

stefpopa commented:
Will merge later if there are no objections.
